moped-editor/src/views/projects/projectView/ProjectPhase/helpers.js

```diff
@@ -0,0 +1,183 @@
+        defaultValues[key] = phase[key];
+      });
+    }
+    return defaultValues;
+  }, [phase]);
+
+/**
+ * Hook which returns an array of subphase options given an input `phase_id`
+ * and an array of `moped_phases` objects
+ */
+export const useSubphases = (phase_id, phases) =>
+  useMemo(
+    () =>
+      phase_id
+        ? phases.find((p) => p.phase_id === phase_id)?.moped_subphases || []
+        : [],
+    [phase_id, phases]
+  );
+
+/**
+ * Hook which returns an object of phase IDs with their name. Taking
+ * the shape of  { [phase_id]: phase_name }
+ */
+export const usePhaseNameLookup = (phases) =>
+  useMemo(
+    () =>
+      phases.reduce(
+        (obj, item) =>
+          Object.assign(obj, {
+            [item.phase_id]: item.phase_name,
+          }),
+        {}
+      ),
+    [phases]
+  );
+
+/**
+ * Hook which returns an object of subphase IDs with their name. Taking
+ * the shape of  { [subphase_id]: subphase_name }
+ */
+export const useSubphaseNameLookup = (subphases) =>
+  useMemo(
+    () =>
+      subphases.reduce(
+        (obj, item) =>
+          Object.assign(obj, {
+            [item.subphase_id]: item.subphase_name,
+          }),
+        {}
+      ),
+    [subphases]
+  );
+
+/**
+ * Hoolk which returns an array of project_phase_ids of the project's current phase(s).
+ * Although only one phase should ever be current, we handle the possibilty that there
+ * are multiple
+ * @param {Array} projectPhases - array of this project's moped_proj_phases
+ * @return {Array} of project_phase_id's of current project phases
+ */
+export const useCurrentProjectPhaseIDs = (projectPhases) =>
+  useMemo(
+    () =>
+      projectPhases
+        ? projectPhases
+            .filter(({ is_current_phase }) => is_current_phase)
+            .map(({ project_phase_id }) => project_phase_id)
+        : [],
+    [projectPhases]
+  );
+
+/**
+ * Hook which returns an array of `moped_proj_phases.project_phase_id`s which
+ * need to have their `is_current` flag cleared.
+ * @param {int} thisProjectPhaseId - the `project_phase_id` that is being edited
+ * @param {bool} isCurrent - if the phase that is being edited is set as the current phase
+ * @param {array} currentProjectPhaseIds - an array of all project_phase_ids that are marked as current.
+ * (this is the output of the useCurrentProjectPhaseIDs hook)
+ * @return {Array} of project_phase_id's which need to set to `is_current` = false
+ */
+export const useCurrentPhaseIdsToClear = (
+  thisProjectPhaseId,
+  isCurrent,
+  currentProjectPhaseIds
+) => {
+  if (!isCurrent) {
+    // nothing to do
+    return [];
+  }
+  // return all project phase IDs except the one we're editing
+  return currentProjectPhaseIds.filter(
+    (projectPhaseId) => projectPhaseId !== thisProjectPhaseId
+  );
+};
+
+export const onSubmitPhase = ({
+  data,
+  mutate,
+  currentPhaseIdsToClear,
+  onSubmitCallback,
+}) => {
+  const { project_phase_id } = data;
+  delete data.project_phase_id;
```

Comment:
I've been recently using destructuring with rest to do this. `const { project_phase_id, ...restOfData } = data;`. It works both ways - just reminded me of code I was writing the other day.

Reply:
and that will remove it? TIL @mddilley